reaktive/src/linuxCommonMain/kotlin/com/badoo/reaktive/utils/Lock.kt

```diff
@@ -0,0 +1,101 @@
+package com.badoo.reaktive.utils
+
+import kotlinx.cinterop.Arena
+import kotlinx.cinterop.CPointer
+import kotlinx.cinterop.alloc
+import kotlinx.cinterop.memScoped
+import kotlinx.cinterop.ptr
+import platform.posix.CLOCK_REALTIME
+import platform.posix.PTHREAD_MUTEX_RECURSIVE
+import platform.posix.clock_gettime
+import platform.posix.pthread_cond_broadcast
+import platform.posix.pthread_cond_destroy
+import platform.posix.pthread_cond_init
+import platform.posix.pthread_cond_t
+import platform.posix.pthread_cond_timedwait
+import platform.posix.pthread_cond_wait
+import platform.posix.pthread_mutex_destroy
+import platform.posix.pthread_mutex_init
+import platform.posix.pthread_mutex_lock
+import platform.posix.pthread_mutex_t
+import platform.posix.pthread_mutex_unlock
+import platform.posix.pthread_mutexattr_destroy
+import platform.posix.pthread_mutexattr_init
+import platform.posix.pthread_mutexattr_settype
+import platform.posix.pthread_mutexattr_t
+import platform.posix.timespec
+
+internal actual class Lock {
+
+    private val arena = Arena()
+    private val attr = arena.alloc<pthread_mutexattr_t>()
+    private val mutex = arena.alloc<pthread_mutex_t>()
+
+    init {
+        pthread_mutexattr_init(attr.ptr)
+        pthread_mutexattr_settype(attr.ptr, PTHREAD_MUTEX_RECURSIVE.toInt())
+        pthread_mutex_init(mutex.ptr, attr.ptr)
+    }
+
+    actual fun acquire() {
+        pthread_mutex_lock(mutex.ptr)
+    }
+
+    actual fun release() {
+        pthread_mutex_unlock(mutex.ptr)
+    }
+
+    actual fun destroy() {
+        pthread_mutex_destroy(mutex.ptr)
+        pthread_mutexattr_destroy(attr.ptr)
+        arena.clear()
+    }
+
+    actual fun newCondition(): Condition = ConditionImpl(mutex.ptr)
+
+    private class ConditionImpl(
+        private val lockPtr: CPointer<pthread_mutex_t>
+    ) : Condition {
+
+        private val arena = Arena()
+        private val cond = arena.alloc<pthread_cond_t>()
+
+        init {
+            pthread_cond_init(cond.ptr, null)
```

Comment:
We should use monotonic time since Linux supports it. Please apply the following patch :-)
```
Index: reaktive/src/nativeCommonMain/kotlin/com/badoo/reaktive/utils/Lock.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- reaktive/src/nativeCommonMain/kotlin/com/badoo/reaktive/utils/Lock.kt	(date 1565812412000)
+++ reaktive/src/nativeCommonMain/kotlin/com/badoo/reaktive/utils/Lock.kt	(date 1565866352000)
@@ -6,14 +6,19 @@
 import kotlinx.cinterop.convert
 import kotlinx.cinterop.memScoped
 import kotlinx.cinterop.ptr
+import platform.posix.CLOCK_MONOTONIC
 import platform.posix.PTHREAD_MUTEX_RECURSIVE
-import platform.posix.gettimeofday
+import platform.posix.clock_gettime
 import platform.posix.pthread_cond_broadcast
 import platform.posix.pthread_cond_destroy
 import platform.posix.pthread_cond_init
 import platform.posix.pthread_cond_t
 import platform.posix.pthread_cond_timedwait
 import platform.posix.pthread_cond_wait
+import platform.posix.pthread_condattr_destroy
+import platform.posix.pthread_condattr_init
+import platform.posix.pthread_condattr_setclock
+import platform.posix.pthread_condattr_t
 import platform.posix.pthread_mutex_destroy
 import platform.posix.pthread_mutex_init
 import platform.posix.pthread_mutex_lock
@@ -23,8 +28,8 @@
 import platform.posix.pthread_mutexattr_init
 import platform.posix.pthread_mutexattr_settype
 import platform.posix.pthread_mutexattr_t
+import platform.posix.time_t
 import platform.posix.timespec
-import platform.posix.timeval
 
 internal actual class Lock {
 
@@ -59,20 +64,20 @@
     ) : Condition {
 
         private val arena = Arena()
+        private val attr = arena.alloc<pthread_condattr_t>()
         private val cond = arena.alloc<pthread_cond_t>()
 
         init {
-            pthread_cond_init(cond.ptr, null)
+            pthread_condattr_init(attr.ptr)
+            pthread_condattr_setclock(attr.ptr, CLOCK_MONOTONIC)
+            pthread_cond_init(cond.ptr, attr.ptr)
         }
 
         override fun await(timeoutNanos: Long) {
             if (timeoutNanos >= 0L) {
                 memScoped {
-                    // iOS does not support pthread_condattr_setclock() nor clock_gettime(), can't use monotonic time
-                    val tv: timeval = alloc { gettimeofday(ptr, null) }
                     val ts: timespec = alloc()
-                    ts.tv_sec = tv.tv_sec
-                    ts.tv_nsec = (tv.tv_usec.convert<Int>() * MICROS_IN_NANOS).convert()
+                    clock_gettime(CLOCK_MONOTONIC, ts.ptr)
                     ts += timeoutNanos
                     pthread_cond_timedwait(cond.ptr, lockPtr, ts.ptr)
                 }
@@ -87,21 +92,20 @@
 
         override fun destroy() {
             pthread_cond_destroy(cond.ptr)
+            pthread_condattr_destroy(attr.ptr)
             arena.clear()
         }
 
         private companion object {
             private const val SECOND_IN_NANOS = 1_000_000_000L
-            private const val MICROS_IN_NANOS = 1000
 
             private operator fun timespec.plusAssign(nanos: Long) {
-                setNanos(tv_sec * SECOND_IN_NANOS + tv_nsec + nanos)
-            }
-
-            private fun timespec.setNanos(nanos: Long) {
-                val secs = nanos / SECOND_IN_NANOS
-                tv_sec = secs.convert()
-                tv_nsec = (nanos - (secs * SECOND_IN_NANOS)).convert()
+                tv_sec += (nanos / SECOND_IN_NANOS).convert<time_t>()
+                tv_nsec += (nanos % SECOND_IN_NANOS).convert<time_t>()
+                if (tv_nsec >= SECOND_IN_NANOS) {
+                    tv_sec += 1
+                    tv_nsec -= SECOND_IN_NANOS.convert<time_t>()
+                }
             }
         }
     }
```

Reply:
`tv_nsec` uses `__syscall_slong_t`.